src/database.rs

```diff
@@ -76,41 +98,90 @@ pub fn get_data_from_db(
             .collect();
 
         if missing_sgs.is_empty() {
-            println!("No missing state groups");
+            // println!("No missing state groups");
             break;
         }
 
         missing_sgs.sort_unstable();
         missing_sgs.dedup();
 
-        println!("Missing {} state groups", missing_sgs.len());
+        // println!("Missing {} state groups", missing_sgs.len());
 
-        let map = get_missing_from_db(&mut client, &missing_sgs);
-        state_group_map.extend(map.into_iter());
+        // find state groups not picked up already and add them to the map
+        let map = get_missing_from_db(&mut client, &missing_sgs, min_state_group, max_group_found);
+        for (k, v) in map.into_iter() {
+            state_group_map.entry(k).or_insert(v);
+        }
     }
 
-    state_group_map
+    (state_group_map, max_group_found)
+}
+
+/// Returns the group ID of the last group to be compressed
+///
+/// This can be saved so that future runs of the compressor only
+/// continue from after this point
+///
+/// # Arguments
+///
+/// * `client`              -   A Postgres client to make requests with
+/// * `room_id`             -   The ID of the room in the database
+/// * `min_state_group`     -   The lower limit (non inclusive) of group id's to compress
+/// * 'groups_to_compress'  -   How many groups to compress
+/// * `max_state_group`     -   The upper bound on what this method can return
+fn find_max_group(
+    client: &mut Client,
+    room_id: &str,
+    min_state_group: Option<i64>,
+    groups_to_compress: Option<i64>,
+    max_state_group: Option<i64>,
+) -> i64 {
+    // Get list of state_id's in a certain room
+    let mut sql = "SELECT id FROM state_groups WHERE room_id = $1".to_string();
+
+    if let Some(max) = max_state_group {
+        sql = format!("{} AND id <= {}", sql, max)
+    }
+
+    // Adds additional constraint if a groups_to_compress has been specified
+    // Then sends query to the datatbase
+    let rows = if let (Some(min), Some(count)) = (min_state_group, groups_to_compress) {
+        let params: Vec<&dyn ToSql> = vec![&room_id, &min, &count];
+        client.query_raw(format!(r"{} AND id > $2 LIMIT $3", sql).as_str(), params)
+    } else {
+        client.query_raw(
+            format!(r"{} ORDER BY id DESC LIMIT 1", sql).as_str(),
```

Comment:
```suggestion
            &format!(r"{} ORDER BY id DESC LIMIT 1", sql),
```

Does that not work?

Reply:
For some reason - no... 

One & gives: `ToStatement` is not implemented for `String`
Two & gives: `ToStatement` is not implemented for `&String`

Reply:
Bleurgh, tedious